### DISCUSSION TITLE
Passing data between routes/components while redirecting

### DISCUSSION THREAD

**Discussion by UzEE** [vercel/next.js#53794] [Help]
### Summary

We're just getting started with Next.js (and the app router) so I apologise if this question is really obvious: we can't figure out a way to pass data to routes or components while doing a redirect.

The `redirect` function from next/navigation only seems to accept a URL and RedirectType arguments. This seems in line with an HTTP redirect but I can't wrap my head around how the following workflow would work with server components:

1. User fills out and submits a form with some data.
2. Data is sent to a handler via maybe a POST request.
3. A corresponding object is created from the data that was submitted on the server.
4. The user is redirected to a "success" page which is composed of some of the data the user submitted.
5. Alternatively, the user can be sent to an "error" page which again would have some of the data the user submitted.

I know how I would go about doing this in a pure client-side SPA but I would imagine it should be doable in a server-render model as well without having to pass data via query params or route-segments, but I can't seem to find any relative APIs or documented patterns supporting this workflow.

**Comment by LimSB-dev** [comment]
You are correct that the redirect function from next/navigation does not support passing data to routes or components. This is because redirects are handled entirely on the server, and there is no way to pass data between the server and the client during a redirect.

There are a few ways to work around this limitation. One way is to use query parameters to pass data to the next page. For example, you could redirect the user to the "success" page with the following code:

```js
const data = {
  name: "John Doe",
  email: "johndoe@example.com",
};

const redirectUrl = `/success?name=${data.name}&email=${data.email}`;

redirect(redirectUrl);
```

The "success" page can then access the data that was passed through the query parameters using the useParams hook.

**Reply by UzEE** [reply]
I am aware that I can pass data via query params, but I'm trying to figure out a pattern for passing data between Server Components directly without exposing it via query params. Not everything can be included in the query itself and it also looks bad from a UX point of view.

We can't use context either since it runs on client, but I find it odd that there isn't a pattern for passing data or state between server components without exposing it to the client side in between. I would've assumed that server components would've made it easier to share data server side without having to rely on a data layer (Redis or a database etc.).
